src/ensembl/doc/architecture/decisions/0018-indexeddb.md

```diff
@@ -0,0 +1,66 @@
+| Year released          | ~ 2009        | ~2012
+| Type of storage        | key-value     | noSQL database
+| Types of stored values | strings       | various, including files
+| Kind of api            | synchronous   | asynchronous
+| Maximum storage size   | 5-10MB        | variable, potentially multiple gigabytes
+| Access from workers    | no            | yes
+
+## Strengths of IndexedDB
+There are several features of IndexedDB that make it a particularly attractive storage option:
+- It is a real database that can have multiple object stores for different data models. As opposed to LocalStorage, it natively supports versioning, which should encourage a more disciplined approach to evolving of our data models (i.e. every breaking change will result in a version bump).
+- It has an asynchronous api, which prevents the blocking of browser's main thread during the saving and the retrieval of data. In contrast, LocalStorage has a synchronous api, which makes it possible for it to block browser's main thread if data transformation to/from strings takes a long time.
+- It has a large storage capacity. Effectively, it can grow as large as there is free space on the disk available to the web browser (roughly up to 80% of free disk space). In practice, this means that it can potentially store gigabytes of data.
+- It is accessible not only from the main thread, but also from workers.
+
+The above characteristics make IndexedDB the preferred option of client-side data storage in modern web development. For example, Google [recommends](https://web.dev/storage-for-the-web) web developers to always pick IndexedDB for the storage of any client-side data other than browser-cached network requests. 
+
+Unfortunately, IndexedDB was released several years before javascript acquired the Promise object for ergonomic control of asynchronous operaitons. Because of that, IndexedDB's api, which is based on event listeners and callbacks, is notoriously clunky. A tiny third-party library `idb` addresses this problem by wrapping IndexedDB into a more convenient promise-based api. It is generally recommended that web developers use IndexedDB with the `idb` library.
+
+## The immediate reason to use IndexedDB for the new Ensembl site
+This ADR is being proposed as we are introducing the Contact Us form. It has been decided that, for usability reasons, user data entered into the form should persist between sessions until the user submits the form, so that, if the user closes the form or accidentally closes the browser tab with the form altogether, the data already entered into the form would remain when the form is re-opened. It has also been decided that the form can include attachments, and the total size of the attachments can be up to 10MB. This last requirement alone makes it impossible for us to store the form data in LocalStorage.
+
+In the future, it may be worth following Google's advice and switch from LocalStorage to IndexedDB entirely.
```

Comment:
Is this for other areas in the new site, such as Browser and Entity Viewer, where we use LocalStorage?

Reply:
Yes.